test/e2e/gameserverallocation_test.go

```diff
@@ -260,6 +263,542 @@ func TestCreateFleetAndGameServerPlayerCapacityAllocation(t *testing.T) {
+
+			gs1, err := framework.AgonesClient.AgonesV1().GameServers(flt.ObjectMeta.Namespace).Get(ctx, gsa.Status.GameServerName, metav1.GetOptions{})
+			require.NoError(t, err)
+			assert.Equal(t, allocated, gs1.Status.State)
+			assert.NotNil(t, gs1.ObjectMeta.Annotations["agones.dev/last-allocated"])
+			require.Equal(t, testCase.wantGameServer, gs1.ObjectMeta.Name)
+
+			// Reset any GameServers in state Allocated -> Ready. Note: This does not reset any changes to Counters.
+			list, err := framework.ListGameServersFromFleet(flt)
+			require.NoError(t, err)
+			for _, gs := range list {
+				if gs.Status.State == ready {
+					continue
+				}
+				gsCopy := gs.DeepCopy()
+				gsCopy.Status.State = ready
+				reqReadyGs, err := framework.AgonesClient.AgonesV1().GameServers(framework.Namespace).Update(ctx, gsCopy, metav1.UpdateOptions{})
+				require.NoError(t, err)
+				require.Equal(t, ready, reqReadyGs.Status.State)
+			}
+		})
+	}
+}
+
+func TestCounterGameServerAllocationActions(t *testing.T) {
+	if !runtime.FeatureEnabled(runtime.FeatureCountsAndLists) {
+		t.SkipNow()
+	}
+	t.Parallel()
+	ctx := context.Background()
+	client := framework.AgonesClient.AgonesV1()
+
+	counters := map[string]agonesv1.CounterStatus{}
+	counters["games"] = agonesv1.CounterStatus{
+		Count:    5,
+		Capacity: 10,
+	}
+
+	flt := defaultFleet(framework.Namespace)
+	flt.Spec.Template.Spec.Counters = counters
+
+	flt, err := client.Fleets(framework.Namespace).Create(ctx, flt.DeepCopy(), metav1.CreateOptions{})
+	assert.NoError(t, err)
+	defer client.Fleets(framework.Namespace).Delete(ctx, flt.ObjectMeta.Name, metav1.DeleteOptions{}) // nolint:errcheck
+	framework.AssertFleetCondition(t, flt, e2e.FleetReadyCount(flt.Spec.Replicas))
+
+	allocated := agonesv1.GameServerStateAllocated
+	ready := agonesv1.GameServerStateReady
+	increment := agonesv1.GameServerPriorityIncrement
+	decrement := agonesv1.GameServerPriorityDecrement
+	zero := int64(0)
+	one := int64(1)
+	five := int64(5)
+	six := int64(6)
+	negativeOne := int64(-1)
+
+	testCases := map[string]struct {
+		gsa          allocationv1.GameServerAllocation
+		wantGsaErr   bool
+		wantCount    *int64
+		wantCapacity *int64
+	}{
+		"increment": {
+			gsa: allocationv1.GameServerAllocation{
+				Spec: allocationv1.GameServerAllocationSpec{
+					Counters: map[string]allocationv1.CounterAction{
+						"games": {
+							Action: &increment,
+							Amount: &one,
+						}}}},
+			wantGsaErr: false,
+			wantCount:  &six,
+		},
+		"decrement": {
+			gsa: allocationv1.GameServerAllocation{
+				Spec: allocationv1.GameServerAllocationSpec{
+					Counters: map[string]allocationv1.CounterAction{
+						"games": {
+							Action: &decrement,
+							Amount: &five,
+						}}}},
+			wantGsaErr: false,
+			wantCount:  &zero,
+		},
+		// TODO: IIRC the SDK with zero out Count if Capacity is set to zero. Do we want this same behavior for GameServerAllocation?
+		"change capacity": {
```

Comment:
Note: this set of "failing" tests aren't doing any validation / assertions currently.

Reply:
Yep, same rules everywhere - if you go below zero, it's still zero.

Or as you asking if we should validate it?

If validation: No, if you want to ensure there is room for a subtraction to occur, then you should add that to the filter. ([design note](https://github.com/googleforgames/agones/issues/2716#:~:text=If%20a%20user%20wants%20to%20ensure%20there%20is%20room%20for%20the%20increment%20or%20decrement%2C%20that%20should%20be%20explicitly%20included%20in%20the%20filter%20options%20(i.e.%20decrement%20by%20one%2C%20but%20filter%20for%20counts%20that%20are%20%3E%200%20so%20that%20there%20is%20something%20to%20decrement).))

Reply:
Asking if we set Capacity to 0, should we also set Count to 0? We do this in the SDK, but we don't have that pattern implemented in GameServerAllocation (Count stays as-is regardless if Capacity is set to less than Count).

Reply:
Also, for CountActions is a action can't be done we drop the action. I.e. if the Count is 5 and the action is to decrement by 6, the CounterAction will simply not happen. Are you saying we should truncate? For SDK we truncate batched requests, however we reject individual requests for `out of range. Count must be within range [0,Capacity]`.

Reply:
> . Are you saying we should truncate

Yes, truncate - since the allocation can't check (or the filter is the check if the end user wants it).

Reply:
My only problem with truncating is that's it's inconsistent with the behavior in the SDK. The SDK will reject any request that tries to increment beyond capacity, or decrement past zero. I don't have any opinion on truncating or not truncating, just that whatever we choose should be consistent across the GSA and SDK.

Reply:
> My only problem with truncating is that's it's inconsistent with the behavior in the SDK

We can't be consistent though, since we don't know at allocation time which GameServer the allocation will be applied to. The SDK is applying an operation to a specific GameServer known ahead of time. I'd argue they are different enough that consistency isn't as strong a requirement.

If the end user wants to ensure that no matter what, there is enough room for the operation to be applied, they can do so with a filter - but the current design doesn't force them to do that if it's not required for their particular use case where truncation is acceptable.

Or to put it another way, if you want it to be consistent (ish?), use a filter expression on your count/list, but we don't require it.